Elements/src/extension.dib

```diff
@@ -0,0 +1,143 @@
+const light = new THREE.AmbientLight(0xaaaaaa);
+scene.add(light);
+
+const directionalLight = new THREE.DirectionalLight(0xaaaaaa, 0.5);
+scene.add(directionalLight);
+
+
+const renderer = new THREE.WebGLRenderer();
+renderer.setSize(width, height);
+document.getElementById('main_DIV_ID').appendChild(renderer.domElement);
+
+camera.position.z = 5;
+
+const controls = new OrbitControls(camera, renderer.domElement);
+
+function base64ToArrayBuffer (base64) {
+    var binary_string = window.atob(base64);
+    var len = binary_string.length;
+    var bytes = new Uint8Array(len);
+    for (var i = 0; i < len; i++) {
+        bytes[i] = binary_string.charCodeAt(i);
+    }
+    return bytes.buffer;
+}
+// adapted from looeee's solution https://discourse.threejs.org/t/camera-zoom-to-fit-object/936/3
+const fitCameraToObject = function (scene, offset = 1.25) {
+    offset = offset || 1.25;
+    let boundingBox = null;
+    scene.traverseVisible((child) => {
+        const objectBox = new THREE.Box3().setFromObject(child)
+        if (boundingBox === null) {
+            boundingBox = objectBox
+        } else {
+            boundingBox = boundingBox.union(objectBox)
+        }
+    })
+    const center = boundingBox.getCenter(new THREE.Vector3());
+    const size = boundingBox.getSize(new THREE.Vector3());
+
+    // get the max side of the bounding box (fits to width OR height as needed )
+    const maxDim = Math.max(size.x, size.y, size.z);
+    const fov = camera.fov * (Math.PI / 180);
+    let cameraZ = Math.abs(maxDim / 4 * Math.tan(fov * 2));
+
+    camera.position.copy(center.clone().add(size.clone().multiplyScalar(offset)))
+
+    const minZ = boundingBox.min.z;
+    const cameraToFarEdge = (minZ < 0) ? -minZ + cameraZ : cameraZ - minZ;
+
+    camera.far = cameraToFarEdge * 3;
+    camera.updateProjectionMatrix();
+
+    if (controls) {
+        // set camera to rotate around center of loaded object
+        controls.target = center;
+        // prevent camera from zooming out far enough to create far plane cutoff
+        controls.maxDistance = cameraToFarEdge * 2;
+        controls.saveState();
+
+    } else {
+        camera.lookAt(center)
+    }
+}
+
+function animate () {
+    requestAnimationFrame(animate);
+    controls.update();
+    renderer.render(scene, camera);
+};
+
+const modelBytes = ""MODEL_BYTES_HERE"";
+
+const gltf = loader.parse(base64ToArrayBuffer(modelBytes), null, (glb) => {
+    loaderScene.add(glb.scene);
+    fitCameraToObject(scene);
+});
+
+
+animate();
+
+</script>
+";
+
+if (KernelInvocationContext.Current is { } currentContext)
+{
+    currentContext.DisplayAs("Add `return model;` at the end of a cell or call `DisplayModel(model, width, height);` to display an Elements model.", "text/markdown");
+}
+
+string GetModelViewerSrc(Model model, double width=400, double height=400) {
+    var gltf = model.ToGlTF();
+    var gltfString = Convert.ToBase64String(gltf, 0, gltf.Length);
+    return viewerSrc
```

Comment:
Could these C# kernel variables could be accessed directly from within the javascript? https://github.com/dotnet/interactive/blob/main/docs/javascript-overview.md#accessing-kernel-data-from-client-side-javascript-code

Reply:
Yes, this works well, and was actually the first version of this I tried:

https://user-images.githubusercontent.com/31935763/147887266-754e461c-4c0b-4485-b76e-be2aa59d0f9a.mp4

Mutating the script string with `Replace` is definitely janky, and I don't think it's the right long-term solution, even if it works fine for right now. But in the end I went this route because I couldn't figure out how to do variable sharing with javascript-in-html generated by the extension... I'm sure there's a way, just haven't gotten that far yet!

Reply:
(The reason I went with an all `#!html` cell instead of a mixture of `#!html` and `#!javascript, which would have enabled variable sharing, is just because I couldn't for the life of me get three.js to load via the requirejs pathway for dependencies in a `#!javascript` cell. Bringing it in via `<script type="module">` in the HTML was the only way I could get threejs going. Again, I'm sure there's a way!)